### PR TITLE
Feature/search in nav

### PR DESCRIPTION
**Pull request by learykara**
Adding one more item to the nav means we need to bump up the breakpoint for the mobile nav again (to ~960px).. is there a way around this? @paulaborowska 

image

### CODE REVIEW

File: src/AppBar/NavLink.jsx [hunk 1]
```diff
@@ -33,4 +35,46 @@ class NavLink extends React.Component {
     }
 }
 
-module.exports = {NavLink};
+class SearchLink extends React.Component {
+    static displayName = "SearchLink";
+    static propTypes = {
```
Comment (ripleyaffect) [inline src/AppBar/NavLink.jsx]: space b/t these

File: src/AppBar/NavLink.jsx [hunk 2]
```diff
@@ -33,4 +35,46 @@ class NavLink extends React.Component {
...
+                    {icon &&
+                        <Icon className="app-nav-icon" name={icon}/>
+                    }
+                    {displayName
+                        && <span className="app-nav-text">{displayName}</span>
+                    }
+                </a>
+                <SearchBar
+                        config={config}
+                        open={open}/>
```
Comment (ripleyaffect) [inline src/AppBar/NavLink.jsx]: What's the difference between passing the prop in and doing `{open && <SearchBar ... />}`?

File: src/AppBar/linkSet.es6 [hunk 3]
```diff
@@ -41,6 +41,11 @@ else {
         menu.push(config.activity);
         main.push(config.workshops);
         main.push(config.library);
+        main.push({
+            displayName: 'Search',
+            icon: 'search',
+            search: true
+        });
```
Comment (ripleyaffect) [inline src/AppBar/linkSet.es6]: this should follow the same pattern as the others

File: src/AppBar/NavLink.jsx [hunk 4]
```diff
@@ -33,4 +35,46 @@ class NavLink extends React.Component {
     }
 }
 
-module.exports = {NavLink};
+class SearchLink extends React.Component {
+    static displayName = "SearchLink";
```
Comment (ripleyaffect) [inline src/AppBar/NavLink.jsx]: Doesn't seem to be used?

File: src/AppBar/search/SearchActions.es6 [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+const superagent = require('superagent');
+
+const SearchActions = Reflux.createActions({
+  getSuggestions: {asyncResult: true}
+});
+
+
+SearchActions.getSuggestions.listen(function(input, config) {
+  superagent.
+    post(`${config.projects.url}/api/search`).
```
Comment (ripleyaffect) [inline src/AppBar/search/SearchActions.es6]: why a post?

File: src/AppBar/search/SearchBar.jsx [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+const {SearchActions} = require('./SearchActions');
+
+const DOWN_ARROW_KEY_CODE = 40;
+const UP_ARROW_KEY_CODE = 38;
+const ENTER_KEY_CODE = 13;
+
+class SearchBar extends React.Component {
+  constructor(props) {
+    super(props);
+    this._handleFormInput = this._handleFormInput.bind(this);
```
Comment (ripleyaffect) [inline src/AppBar/search/SearchBar.jsx]: arrow syntax to avoid this

File: src/AppBar/search/SearchBar.jsx [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+      selectedSuggestionIdx: -1,
+      suggestedEnd: '',
+      suggestions: []
+    }
+  }
+
+  componentDidMount() {
+    SearchActions.getSuggestions.completed.listen(this._onGetSuggestionsCompleted);
+    React.findDOMNode(this.refs.input).
+      addEventListener('keydown', e => this._handleKeyDown(e));
```
Comment (ripleyaffect) [inline src/AppBar/search/SearchBar.jsx]: this arrow function doesn't really do anything, you can just use `addEventListener('keydown', this._handleKeyDown);`

File: src/AppBar/search/SearchBar.jsx [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+            onSubmit={this._handleSubmitForm}>
+          <h4 className="search-heading">What would you like to learn?</h4>
+          <Icon name="search" className="icon-search"/>
+          <input
+              className="search-input"
+              ref="input"
+              type="text"
+              onChange={this._handleFormInput}
+              value={searchTerm}/>
+          {searchTerm.length &&
```
Comment (ripleyaffect) [inline src/AppBar/search/SearchBar.jsx]: `!!searchTerm.length`?

File: src/AppBar/search/SearchBar.jsx [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+                name="close"
+                className="icon-close"
+                onClick={this._handleClearInput}/>
+          }
+        </form>
+        {!!suggestions.length &&
+          <div className="search-suggestions">
+            {suggestions.map((term, idx) => (
+              <div
+                  key={idx}
```
Comment (ripleyaffect) [inline src/AppBar/search/SearchBar.jsx]: I would actually use the term as the key, it's unique

File: src/AppBar/search/SearchBar.jsx [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+                  className={cx(
+                    "suggestion-item",
+                    {"suggestion-item__active": idx === selectedSuggestionIdx})}
+                  onClick={e => this._handleSubmitForm(e)}
```
Comment (ripleyaffect) [inline src/AppBar/search/SearchBar.jsx]: the arrow function isn't needed here either

File: src/AppBar/search/SearchBar.jsx [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+                  onMouseEnter={e => this._handleSuggestionMouseEnter(idx)}>
```
Comment (ripleyaffect) [inline src/AppBar/search/SearchBar.jsx]: Arrow function is useful here because you're changing the param to the handler

**Comment by ripleyaffect** [comment]
few comments, mostly lgtm :+1: Are we releasing with a feature flag?

File: src/AppBar/NavLink.jsx (hunk 2 above)
Comment (learykara) [inline src/AppBar/NavLink.jsx]: I have it fading in/out, which isn't possible if it's being removed from the dom when it's not open

File: src/AppBar/NavLink.jsx (hunk 4 above)
Comment (learykara) [inline src/AppBar/NavLink.jsx]: this is for the thinkful-ui demo page

File: src/AppBar/search/SearchActions.es6 (hunk 5 above)
Comment (learykara) [inline src/AppBar/search/SearchActions.es6]: hmm. not sure. will change

Comment (learykara) [inline src/AppBar/search/SearchActions.es6]: ah, because I'm sending the term as data. no good?

File: src/AppBar/NavLink.jsx (hunk 2 above)
Comment (ripleyaffect) [inline src/AppBar/NavLink.jsx]: [LIES!](https://facebook.github.io/react/docs/animation.html#animating-one-or-zero-items)

File: src/AppBar/search/SearchActions.es6 (hunk 5 above)
Comment (ripleyaffect) [inline src/AppBar/search/SearchActions.es6]: probably should just be an arg

File: src/AppBar/NavLink.jsx (hunk 2 above)
Comment (ripleyaffect) [inline src/AppBar/NavLink.jsx]: That's fine, btw. Just so you're aware of the transition group option.

Comment (learykara) [inline src/AppBar/NavLink.jsx]: UGH YOU WIN. I'll use that going forward

**Pull request merged**